src/sema/eval.rs

```diff
@@ -196,3 +196,324 @@ pub fn eval_const_rational(
+                    None
+                } else {
+                    Some(Expression::NumberLiteral(*loc, ty.clone(), left % right))
+                }
+            } else {
+                None
+            }
+        }
+        Expression::Power(loc, ty, _, left, right) => {
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (
+                Some(Expression::NumberLiteral(_, _, left)),
+                Some(Expression::NumberLiteral(right_loc, _, right)),
+            ) = (&left, &right)
+            {
+                if let Some(diagnostic) = overflow_check(right, &Type::Uint(16), right_loc) {
+                    ns.diagnostics.push(diagnostic);
+                    None
+                } else {
+                    Some(Expression::NumberLiteral(
+                        *loc,
+                        ty.clone(),
+                        left.pow(right.to_u16().unwrap().into()),
+                    ))
+                }
+            } else {
+                None
+            }
+        }
+        Expression::ShiftLeft(loc, ty, left, right) => {
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (
+                Some(Expression::NumberLiteral(_, _, left)),
+                Some(Expression::NumberLiteral(right_loc, _, right)),
+            ) = (&left, &right)
+            {
+                if right.sign() == Sign::Minus {
+                    ns.diagnostics.push(Diagnostic::error(
+                        *right_loc,
+                        format!("left shift by {} is not possible", right),
+                    ));
+                    None
+                } else {
+                    if right >= &BigInt::from(left.bits()) {
+                        ns.diagnostics.push(Diagnostic::warning(
+                            *right_loc,
+                            format!("left shift by {} may overflow the final result", right),
+                        ));
+                    }
+
+                    Some(Expression::NumberLiteral(
+                        *loc,
+                        ty.clone(),
+                        left.shl(right.to_u64().unwrap()),
+                    ))
+                }
+            } else {
+                None
+            }
+        }
+
+        Expression::ShiftRight(loc, ty, left, right, _) => {
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (
+                Some(Expression::NumberLiteral(_, _, left)),
+                Some(Expression::NumberLiteral(_, _, right)),
+            ) = (&left, &right)
+            {
+                Some(Expression::NumberLiteral(
+                    *loc,
+                    ty.clone(),
+                    left.shr(right.to_u64().unwrap()),
```

Comment:
We need the same for shift right as you did for shift left. Right now you can make the compiler panic with:

```solidity
int x = 1 >> 14676683207225698178084221555689649093015162623576402558976;
```